src/neighborsearch/filtered.jl

```diff
@@ -0,0 +1,146 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    FilteredSearch(method, nmax=Inf; maxpercategory, maxpersector, order=Euclidean())
+
+A method for searching at most `nmax` neighbors using `method`. Extra
+constraints available: `maxpercategory` and `maxpersector`. The priority is
+given to the nearest neighbor using `order`.
+
+## Max per category
+
+It can be a `NamedTuple` or `Dict`, where the first element is the property name
+and the second defines the max neighbor per available categories of the property.
+In the example below, no more than 2 neighbors from the same hole are selected.
+
+`FilteredSearch(method, maxpercategory = (holeid = 2,))`
+[ASCII illustration]
+
+## Max per sector
+
+Only allow a max number of neighbors inside each sector. The sectors are the
+cartesian quadrants (2-D) or octants (3-D). If the neighborhood is an `Ellipsoid`,
+the sectors will match the rotated quadrants/octants.
+
+`FilteredSearch(method, maxpersector = 2)`
+[ASCII illustration]
+"""
+struct FilteredSearch{M<:NeighborSearchMethod} <: BoundedNeighborSearchMethod
+  method::M
+  nmax::Int
+  maxpercategory
+  maxpersector
+  order
+end
+
+FilteredSearch(method::M, nmax=0; maxpercategory=nothing,
+               maxpersector=nothing, order=Euclidean()) where {M} =
+  FilteredSearch{M}(method, nmax, maxpercategory, maxpersector, order)
+
+maxneighbors(method::FilteredSearch) = method.nmax
+
+function search!(neighbors, pₒ::Point, method::FilteredSearch; mask=nothing)
+  meth = method.method
+  obj  = meth.domain
+  inds = search(pₒ, meth, mask=mask)
+  nmax = method.nmax == 0 ? Inf : method.nmax
+
+  # get distances and give priority to closest neighbors if necessary
+  if !(meth isa KNearestSearch)
+    indcoords = reduce(hcat, coordinates.([centroid(obj, s) for s in inds]))
+    dists     = colwise(method.order, coordinates(pₒ), indcoords)
```

Comment:
I think we can avoid this major allocation of points and can instead calculate the distances with a single list comprehension with the simple `evaluate` function:

```julia
dists = [evaluate(method.order, coordinates(p0), coordinates(centroid(obj, ind))) for ind in inds]
```
```

Reply:
Ok, I'll change it. The colwise seems to be faster only if order = Mahalanobis (e.g. Ellipsoid.metric)